docs/enhancements/add-external-providers.md

```diff
@@ -0,0 +1,187 @@
+---
+title: Add-Providers-Support
+authors:
+  - "@Gal-Zaidman"
+creation-date: 2021-07-25
+last-updated: 2021-07-25
+---
+
+# Add Providers Support
+
+## Summary
+
+Currently, the assisted installer supports Bare-Metal installations only.
+Openshift can be installed on various providers such as:
+- Red Hat Virtualization.
+- OpenStack.
+- vSphere.
+- AWS.
+- GCP.
+- Azure.
+
+This enhancement will define a clear interface to extend the assisted installer and allow
+provider installations.
+
+## Motivation
+
+Assisted Installer can be a great tool to perform UPI (User Provisioned) installation on external providers,
+providing the user with clear UI to see its provisioned machines joining the cluster, and lowering installation
+resources requirements.
+
+Assisted Installer is also ideal for mixed env configurations.
+** The RHV provider supports installing a mixed env where the Control-Plain (Master) nodes are deployed on top of RHV
+and Compute (Worker) nodes are Bare-Metal nodes.
```

Comment:
How do we plan to handle the expansion of the test matrix here?  Would we support every provider for single-provider and mixed-provider, or only $provider and baremetal (as is possible today due to the hard-coded `baremetal` platform in AI)

A common requirement is to deploy controlplane nodes on some kind of VM platform, then deploy worker hosts on BM (potentially in an automated way using Metal3), but if we enable it via this AI API, how will we ensure all-the-combinations actually work?

So far we've deferred adding a generic [CBO](https://github.com/openshift/cluster-baremetal-operator) API due to this test-matrix concern - in particular network configuration of controlplane hosts to enable PXE may be different depending on the controlplane provider, etc?

I also wonder about the customer experience vs IPI - e.g in this case you deploy on RHV and have to make a day-1 choice to only ever deploy either RHV or baremetal workers.  So probably customers will then ask for a combination of both, which isn't possible via the current MAPI integration in OCP /cc @JoelSpeed

Reply:
> How do we plan to handle the expansion of the test matrix here? Would we support every provider for single-provider and mixed-provider, or only $provider and baremetal (as is possible today due to the hard-coded `baremetal` platform in AI)

Amm not sure what test matrix you are referring to, to my knowledge RHV is the only provider that allows such mix env, meaning have platform integration (the ovirt provider and csi driver) but still have some BM workers and some oVirt workers.
So I don't think it would affect other providers.

> A common requirement is to deploy controlplane nodes on some kind of VM platform, then deploy worker hosts on BM (potentially in an automated way using Metal3), but if we enable it via this AI API, how will we ensure all-the-combinations actually work?

I think what you mentioned above is archived with setting platform to none, which means no provider integration and in that case we support "Installing a cluster on any platform" flow.
In RHV we allow a unique case in which we have full-provider integration but allow for non RHV nodes.

> So far we've deferred adding a generic [CBO](https://github.com/openshift/cluster-baremetal-operator) API to enable Metal3 on any platform due to this test-matrix concern - in particular network configuration of controlplane hosts to enable PXE may be different depending on the controlplane provider, etc?
> 
> I also wonder about the customer experience vs IPI - e.g in this case you deploy on RHV and have to make a day-1 choice to only ever deploy either RHV or baremetal workers. So probably customers will then ask for a combination of both, which isn't possible via the current MAPI integration in OCP /cc @JoelSpeed

On RHV this is possible, due to the fact that RHV provider is a part of the machine-controller.
This is the [PR which allowed it](https://github.com/openshift/cluster-api-provider-ovirt/pull/95) and this is the [BZ](https://bugzilla.redhat.com/show_bug.cgi?id=1929702)

Reply:
It is something that apparently has become pretty common on RHV and a good use-case for the assisted installer for us.
You can do it with IPI or UPI it doesn't matter, and using the AI with platform set to ovirt will allow users to see their bare metals joining the OCP on RHV cluster

Reply:
> Amm not sure what test matrix you are referring to

I mean which combinations will we actually test and expect to support?  Adding flexibility to mix platforms means we'll have to test the combinations we expect to support, and find some way (at least via docs) to communicate the supported combinations to users.

Reply:
I think specific platforms or combinations thereof are out of scope here.  This is talking about creating some framework for easily adding platforms.  Before any platform is added, there must be approval from PM, QE, etc. so that it will be supported.

Reply:
@avishayt - Let's not confuse this open source contribution with what we'll end up supporting in our downstream product or service offerings. The former does not require any approval apart from the maintainers of this project - if it doesn't break other flows, if it's aligned with the project goals, if it's technically sound, it should go in.

Reply:
Although Assisted Installer is open source, there's no community around it yet and nobody else is using it.  There's no reason to include code which we won't test and won't use.

Reply:
@hardys the only added combinations here are OCP on oVirt combinations:
- with mixed env (since only oVIrt is supporting it).
- All nodes are oVirt vms.

@avishayt you are right that to get it downstream we will need PM and QE approval, and we are working on that front.
Not sure how assisted installer U/S and D/S work is being handled and if it is possible to merge something into the assisted service without it going into D/S... but this complication may not be required.
Since a lot of the vSphere work is in, we can get the interface and port the vSphere work to use it and after that add oVirt provider (I assume that by then we will have the approval we need) and this should be an easier part.

Reply:
> Let's not confuse this open source contribution with what we'll end up supporting in our downstream product or service offerings. The former does not require any approval apart from the maintainers of this project - if it doesn't break other flows, if it's aligned with the project goals, if it's technically sound, it should go in.

@Gal-Zaidman as is typical with OSS projects, a key criteria for merging something big and new, like support for a new platform, is having people in-place who are committed to maintaining it. If we expect RH engineers to maintain support for a platform in this project upstream, then we need RH alignment on committing that engineering time. The same would apply to any other contributor proposing to add a big new feature; they would need to demonstrate that there are people able and willing to maintain it in the upstream project.

You are correct that the maintainers upstream can merge what they see fit, and that doesn't have to directly impact downstream. But responsible upstream maintainers will ensure that a contribution has sufficient ongoing support.

Reply:
> On RHV this is possible, due to the fact that RHV provider is a part of the machine-controller.
This is the PR which allowed it and this is the BZ

I don't think this is what Steve was referring to in his comment.

 MAO deploys machine controllers based on the value of the `Infrastructure` resource `status.platformStatus.type`. When installing, the customer will set this to the main topology for the cluster, I expect to match the control plane. In the case of an RHV control plane, we expect the oVirt Machine Controller to be deployed, but not the baremetal machine controller.

So any Machine that is created, is expected to be an oVirt machine. A customer might expect to be able to create an oVirt machine and a baremetal machine, but the oVirt controller is the only one running, and it won't understand the baremetal machines so will error on them constantly.

Machine API isn't designed to work multi platform today, so I don't think there's a way we can deploy both oVirt and baremetal machine controllers simultaneously and expect them to work

Reply:
I think you meant to tag @mykaul
> @Gal-Zaidman as is typical with OSS projects, a key criteria for merging something big and new, like support for a new platform, is having people in-place who are committed to maintaining it. If we expect RH engineers to maintain support for a platform in this project upstream, then we need RH alignment on committing that engineering time. The same would apply to any other contributor proposing to add a big new feature; they would need to demonstrate that there are people able and willing to maintain it in the upstream project.
> 

Yes I agree, but this is something that you already have since the OCP on RHV team agreed to maintain the code and CI for the effort

> You are correct that the maintainers upstream can merge what they see fit, and that doesn't have to directly impact downstream. But responsible upstream maintainers will ensure that a contribution has sufficient ongoing support.

Reply:
We'd be more than happy to maintain it. This is applicable for both RHV and OSP.

Reply:
> > On RHV this is possible, due to the fact that RHV provider is a part of the machine-controller.
> > This is the PR which allowed it and this is the BZ
> 
> I don't think this is what Steve was referring to in his comment.
> 
> MAO deploys machine controllers based on the value of the `Infrastructure` resource `status.platformStatus.type`. When installing, the customer will set this to the main topology for the cluster, I expect to match the control plane. In the case of an RHV control plane, we expect the oVirt Machine Controller to be deployed, but not the baremetal machine controller.
> 
That is correct in the mix-env deployment the only machine controller is the oVirt machine controller, meaning that there will be only oVirt machines, but there can be **nodes** that are not oVirt and that is fine.

> So any Machine that is created, is expected to be an oVirt machine.
+1

> A customer might expect to be able to create an oVirt machine and a baremetal machine, but the oVirt controller is the only one running, and it won't understand the baremetal machines so will error on them constantly.
> 
Yes but a user can have a baremetal nodes and the controller will not care.
 
> Machine API isn't designed to work multi platform today, so I don't think there's a way we can deploy both oVirt and baremetal machine controllers simultaneously and expect them to work

Reply:
I'm switching the focus of this enhancement to be specific to the provider interface and not to add the oVirt provider.
As discussed on other threads as well, the ovirt provider addition is more of a result and maybe will be the first example of this interface implementation